File: frontend/types/spellcheck.ts
export type SpellCheckCandidateResponse = {
  word: string;
  distance: number;
  frequency: number;
};

export type SpellCheckWordResponse = {
  word: string;
  is_valid: boolean;
  suggestions: SpellCheckCandidateResponse[];
};

export type SpellCheckIssueResponse = {
  word: string;
  start: number;
  end: number;
  suggestions: SpellCheckCandidateResponse[];
};

export type SpellCheckTextResponse = {
  issues: SpellCheckIssueResponse[];
  issue_count: number;
};

File: backend/app/api/schemas/spellcheck.py
from pydantic import BaseModel, Field


class SpellCheckCandidateResponse(BaseModel):
    word: str
    distance: int = Field(ge=0)
    frequency: int = Field(ge=0)


class SpellCheckWordResponse(BaseModel):
    word: str
    is_valid: bool
    suggestions: list[SpellCheckCandidateResponse]


class SpellCheckIssueResponse(BaseModel):
    word: str
    start: int = Field(ge=0)
    end: int = Field(ge=0)
    suggestions: list[SpellCheckCandidateResponse]


class SpellCheckTextRequest(BaseModel):
    text: str = Field(default="", max_length=20000)


class SpellCheckTextResponse(BaseModel):
    issues: list[SpellCheckIssueResponse]
    issue_count: int = Field(ge=0)


class LevenshteinTableResponse(BaseModel):
    source: str
    target: str
    table: list[list[int]]
    distance: int = Field(ge=0)